vdj_ann/src/transcript.rs

```diff
@@ -5,29 +5,49 @@
 use crate::annotate::get_cdr3_using_ann;
 use crate::refx::RefData;
 use amino::{have_start, have_stop};
-use debruijn::{dna_string::DnaString, kmer::Kmer20, Mer, Vmer};
+use debruijn::dna_string::DnaString;
+use debruijn::kmer::Kmer20;
+use debruijn::{Mer, Vmer};
 use hyperbase::Hyper;
 use io_utils::fwriteln;
 use kmer_lookup::make_kmer_lookup_20_single;
-use std::{cmp::max, io::prelude::*};
+use std::cmp::max;
+use std::io::prelude::*;
 use vector_utils::{lower_bound1_3, unique_sort};
 
 // ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
 // TEST FOR VALID VDJ SEQUENCE
 // ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
 
+#[derive(Debug, PartialEq, Ord, PartialOrd, Eq)]
+pub enum UnproductiveContigCause {
+    NoCdr3,
+    Misordered,
+    NotFull,
+    TooLarge,
+}
+
+#[derive(Debug)]
+pub struct ContigStatus {
+    pub productive: bool,
+    pub unproductive_cause: Vec<UnproductiveContigCause>,
+}
+
 pub fn is_valid(
     b: &DnaString,
     refdata: &RefData,
     ann: &[(i32, i32, i32, i32, i32)],
     logme: bool,
     log: &mut Vec<u8>,
     is_gd: Option<bool>,
-) -> bool {
+) -> ContigStatus {
     // Unwrap gamma/delta mode flag
     let gd_mode = is_gd.unwrap_or(false);
     let refs = &refdata.refs;
     let rheaders = &refdata.rheaders;
+    let mut ret_vec = Vec::new();
+    let mut never_full = true;
+    // two passes, one for light chains and one for heavy chains
     for pass in 0..2 {
         let mut m = "A";
         if pass == 1 {
```

Comment:
This `ann` tuple is all over the vdj code base and it makes everything so confusing and hard to follow. Converting this into a struct is a lot more work but at-least in this function if we could rename the fields to something more intuitive it would really help. I think its defined here https://github.com/10XGenomics/rust-toolbox/blob/03aa9d3ebaf640afb871c9d59b9ef63f426a309e/vdj_ann/src/annotate.rs#L1098